Check database version in migrate command

ausi opened the pull request:
As discussed in https://github.com/contao/contao/pull/4158#issuecomment-1169685620 (and on Slack)

This only fails if the configured server version causes the wrong doctrine platform to be used. 
So if the actual version is `8.0.29` but `8.0.28` is configured no error occurs.
But if `5.7.0` is configured it would show an error message like: 
`Wrong database version, please set it to "8.0.29". Expected "Doctrine\DBAL\Platforms\MySQL80Platform" was "Doctrine\DBAL\Platforms\MySQLPlatform".`

ausi committed:
Check database version in migrate command

ausi committed:
Use get_class instead of ::class

ausi committed:
Fix test

fritzmg reviewed (dismissed):
Tested and looks good to me. (However, I don't find the _"Expected ..."_ part of the error message particularly useful.)

fritzmg commented:
Btw. shouldn't this go into `4.9`? It would be just as helpful there (as for performance you can also define the database version in your `parameters.yaml`, `DATABASE_URL` environment variable or `config.yaml` there).

ausi commented:
> Btw. shouldn't this go into `4.9`?

Because of several merge conflicts that would arise I created a separate PR for 4.9 now: #5101

aschempp reviewed (changes requested):
If the output format is `ndjson`, this should not throw an exception (result in a `error` type message) but generate a `problem` (see https://github.com/contao/contao/pull/5065). Otherwise the Contao Manager cannot see if there is a connection problem/runtime exception or a database-level error.

ausi committed:
Report wrong database version as 'database-error'

# Conflicts:
#	core-bundle/src/Command/MigrateCommand.php
#	core-bundle/tests/Command/MigrateCommandTest.php

ausi committed:
Merge branch origin/4.13 into fix/database-version-check

ausi commented:
> If the output format is `ndjson`, this should not throw an exception

Done in 6340b8642a842f653645fd7fc166be1d9fe6400e

ausi commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -537,4 +546,40 @@ private function compileTableWarnings(Table $table): array
+        }
+
+        if ($asJson) {
+            $this->writeNdjson('database-error', ['message' => $message]);
```
The correct name needs to be resolved in https://github.com/contao/contao/pull/5065#discussion_r940079950 first

ausi replied:
```suggestion
            $this->writeNdjson('problem', ['message' => $message]);
```

ausi committed:
Fix test

ausi committed:
Update core-bundle/src/Command/MigrateCommand.php

ausi commented on core-bundle/tests/Command/MigrateCommandTest.php:
```diff
@@ -429,6 +434,65 @@ public function testDoesAbortOnFatalError(string $format): void
+        if ('ndjson' === $format) {
+            $json = $this->jsonArrayFromNdjson($display)[0];
+
+            $this->assertSame('database-error', $json['type']);
```
```suggestion
            $this->assertSame('problem', $json['type']);
```

ausi committed:
Update core-bundle/tests/Command/MigrateCommandTest.php

aschempp reviewed (dismissed):
not sure why we need this for upstream merge, but code looks good 🙃

leofeyer commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -537,4 +546,40 @@ private function compileTableWarnings(Table $table): array
+    {
+        // TODO: Find a replacement for getWrappedConnection() once doctrine/dbal 4.0 is released
+        $driverConnection = $this->connection->getWrappedConnection();
+        $currentPlatform = \get_class($this->connection->getDatabasePlatform());
```
<img width="903" alt="" src="https://user-images.githubusercontent.com/1192057/184342447-65c15db3-485c-46de-b931-bd28b72ffbcf.png">

Is this going to be a problem?

ausi replied:
I don’t think so, see https://github.com/contao/contao/pull/5133

leofeyer replied:
See https://github.com/ausi/contao-contao/pull/4 😄

ausi replied:
Merged.

leofeyer committed:
Work around the PhpStorm issue

leofeyer committed:
Merge branch '4.13' into fix/database-version-check

leofeyer committed:
CS

leofeyer commented:
Thank you @ausi.
